### DISCUSSION TITLE
Dynamic egui::Window rendering doesn't work with WASM (eframe_template)

### DISCUSSION THREAD

**Discussion by deft01** [emilk/egui#2556] [Q&A] [answered]
I made this code, it use eframe_template repo. It compiles and work well with linux as target.

`selector()` function is executed each frames.
`SelectorWindow` objects are stored in the App, describes the windows `egui::Window` that must be rendered (data storage for windows).
If there are ten `SelectorWindow` objects in the selector**s** vec it render ten windows on the screen (c.f. screenshot at the end of this post).

**Now I don't know why this is not working with WASM as target. I compile with trunk as mentionned in the eframe_template tutorial.**
When clicked, the button has no effect, no errors in the JS console. Not even a blinking Window rendered.

```rust
pub struct SelectorWindow {
    pub is_visible: bool,
    pub title: String,
}

pub fn selector(mut _self:&mut TemplateApp, ui: &mut egui::Ui, ctx: &egui::Context) {
    for selector in _self.selectors.iter_mut() {
        egui::Window::new(&selector.title)
            .resizable(true)
            .show(ctx, |ui| {
                ui.label("Selector");
            });
    }
}
```

Screenshot
![image](https://user-images.githubusercontent.com/10063471/211173568-2c2399c4-acc5-4158-b099-2ce9f0e9eedd.png)

**Comment by coderedart** [comment]
EDIT: answered on discord. the selector windows code was platform gated to `not(taget_arch = "wasm32")`, so it only ran on desktop

> When clicked, the button has no effect,
what button?

Also, is a fresh template repo working on wasm? post a screenshot of what is actually working on browser.

check your devtools/console just to see if there's any logging at all. as you can see at sample https://emilk.github.io/eframe_template/ , there's some logging by eframe about webgl stuff before starting your app. so, if there's no logs at all, it means we never reached the startup stage. 

check if trunk actually compiled successfully? when you use `trunk serve`, it will serve even if build failed. try using `trunk build` and make sure that build is successful. or use `cargo build --target=wasm32-unknown-unknown`.

If instead your problem is just the selector windows part not working. you might be dealing with the cache effect of browser. sometimes, browser will load the cached wasm/js files, and show you the older builds. use `ctrl + F5` to force refresh and load the latest build files.

**Reply by deft01** [reply]
I've already done all the checks :
- Everything is compiling well for both targets (WASM and x86). I also rebuilt.
- The button spawning the window is not working on WASM.
- I confirm I reload the cache everytime (`ctrl + F5`, `ctrl + maj + F5`).

Demo of the feature working well for x86: 

https://user-images.githubusercontent.com/10063471/211188988-fd8ecf18-2b2f-4577-80e7-1ea283ba795b.mp4

Demo of the same code not working with WASM : 

https://user-images.githubusercontent.com/10063471/211189118-a263a1d5-57f2-4b72-8790-bcc76d325a9c.mp4

The content of the console :
The first error comes from a firefox plugin I use and is not related to egui. 
![image](https://user-images.githubusercontent.com/10063471/211189236-7402f054-922f-49bc-989e-90e88f4e31e8.png)

Same console in chrome without any errors : 
![image](https://user-images.githubusercontent.com/10063471/211189286-322d6d84-0450-40f7-8c36-15db71864f65.png)

**Comment by deft01** [comment, marked as answer]
Fixed thanks to @coderedart.
The problem was that the rendering of the windows wasn't setup on a Panel not rendered with wasm.
I didn't pay attention to it because I wasn't aware the default TopPanel with eframe_template wasn't displayed in WASM.

```rust
[cfg(not(target_arch = "wasm32"))] // desactivate this Panel for WASM
gui::TopBottomPanel::top("top_panel").show(ctx, |ui| {
```

Thanks again @coderedart
